src/main/java/com/aws/greengrass/deployment/errorcode/DeploymentErrorCodeUtils.java

```diff
@@ -0,0 +1,225 @@
+                                                                       Map<String, DeploymentErrorCode> errorContext) {
+        errorCodeSet.add(CLOUD_SERVICE_ERROR);
+        if (e instanceof ResourceNotFoundException) {
+            errorCodeSet.add(RESOURCE_NOT_FOUND);
+        } else if (e instanceof AccessDeniedException) {
+            errorCodeSet.add(ACCESS_DENIED);
+        } else if (e instanceof ValidationException) {
+            errorCodeSet.add(BAD_REQUEST);
+        } else if (e instanceof ThrottlingException) {
+            errorCodeSet.add(THROTTLING_ERROR);
+        } else if (e instanceof ConflictException) {
+            errorCodeSet.add(CONFLICTED_REQUEST);
+        } else if (e instanceof InternalServerException) {
+            errorCodeSet.add(SERVER_ERROR);
+        }
+        if (errorContext.containsKey(GreengrassV2DataException.class.getSimpleName())) {
+            errorCodeSet.add(errorContext.get(GreengrassV2DataException.class.getSimpleName()));
+        }
+    }
+
+
+    private static void collectErrorCodesFromS3Exception(Set<DeploymentErrorCode> errorCodeSet, S3Exception e,
+                                                         Map<String, DeploymentErrorCode> errorContext) {
+        errorCodeSet.add(S3_ERROR);
+        int s3StatusCode = e.statusCode();
+        if (s3StatusCode == HttpStatusCode.NOT_FOUND) {
+            errorCodeSet.add(S3_RESOURCE_NOT_FOUND);
+        } else if (s3StatusCode == HttpStatusCode.FORBIDDEN) {
+            errorCodeSet.add(S3_ACCESS_DENIED);
+        } else if (s3StatusCode >= HttpStatusCode.BAD_REQUEST && s3StatusCode < HttpStatusCode.INTERNAL_SERVER_ERROR) {
+            errorCodeSet.add(S3_BAD_REQUEST);
+        } else if (s3StatusCode >= HttpStatusCode.INTERNAL_SERVER_ERROR) {
+            errorCodeSet.add(S3_SERVER_ERROR);
+        }
+
+        if (errorContext.containsKey(S3Exception.class.getSimpleName())) {
+            errorCodeSet.add(errorContext.get(S3Exception.class.getSimpleName()));
+        }
+    }
+
+    private static void collectErrorCodesFromServiceLoadException(Set<DeploymentErrorCode> errorCodeSet,
+                                                                  Map<String, DeploymentErrorCode> errorContext) {
+        errorCodeSet.add(LOCATE_INSTALLED_COMPONENT_ERROR);
+        if (errorContext.containsKey(ServiceLoadException.class.getSimpleName())) {
+            errorCodeSet.add(errorContext.get(ServiceLoadException.class.getSimpleName()));
+        }
+    }
+
+    /**
+     * Helper function to inline add an error code to the head of list.
+     *
+     * @param errorCodeList the destination list
+     * @param errorCode     error code to be added to the head
+     * @return the updated list
+     */
+    public static List<DeploymentErrorCode> inlineAddToFront(List<DeploymentErrorCode> errorCodeList,
+                                                             DeploymentErrorCode errorCode) {
+        errorCodeList.add(0, errorCode);
```

Comment:
warning that you cannot use this for `Arrays.asList` as they're unmodifiable. You may way to handle that better here, and return a list which you can modify.

Reply:
Yea I was planning to wrap a `new ArrayList<>` around `Arrays.asList` before it passes into this function, but I agree maybe I should just return a new list here instead.

Reply:
It is fine if you're modifying a modifiable list, just maybe add a check for an unmodifyable list too and handle that case.

Reply:
Updated to always use a modifiable list. Also added a check for unsupported operation.